File: impl/python/pairl/validate.py
"""PAIRL v1.5 validation rules (V1–V12) operating on a parsed Message."""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from .core import COLUMNAR_FORBIDDEN, Message, split_fields

_REF = re.compile(r"^ref:[A-Za-z0-9_-]+:[^\s]+$")
_SLOC = re.compile(r"^@[A-Za-z0-9]{1,8}(?:#[A-Za-z0-9_-]{1,8})?$")
_BARE_DEP = re.compile(r"^[A-Za-z0-9]{1,8}(?:#[A-Za-z0-9_-]{1,8})?$")
_BUDGET = re.compile(r"^([0-9]+(?:\.[0-9]+)?)([A-Za-z]{1,16})$")
_NUMERIC_INTENT_KEYS = {"l", "m"}


def is_valid_ref(v: str) -> bool:
    if not v.startswith("ref:") or " " in v:
        return False
    main = v.split("#", 1)[0]
    return len(main.split(":")) >= 3 and bool(_REF.match(main))


def is_sloc_ref(v: str) -> bool:
    return bool(_SLOC.match(v))


@dataclass
class Result:
    errors: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)

    @property
    def valid(self) -> bool:
        return not self.errors


def validate(msg: Message, strict: bool = False) -> Result:
    res = Result()
    res.errors.extend(msg.errors)

    # Required headers
    for h in ("v", "ts"):
        if h not in msg.headers:
            res.errors.append(f"missing required header: @{h}")
    if "id" not in msg.headers and "mid" not in msg.headers:
        res.errors.append("missing required header: @id (or legacy @mid)")

    _v1_no_new_facts(msg, res, strict)
    _v2_evidence(msg, res)
    _v3_refs(msg, res)
    _v6_rid_unique(msg, res)
    _v8_budget(msg, res)
    _v9_tool_chain(msg, res, strict)
    _v11_turn_markers(msg, res)
    _v12_columnar(msg, res)
    return res


def _has_rule(msg: Message, name: str) -> bool:
    return any(r.kind == "rule" and r.kv.get(name) == "true" for r in msg.records)


def _v1_no_new_facts(msg: Message, res: Result, strict: bool) -> None:
    enforce = strict and _has_rule(msg, "no_new_facts")
    for r in msg.records:
        if r.kind != "intent":
            continue
        for k, v in r.kv.items():
            if re.search(r"https?://", v):
                _emit(res, enforce, f"V1: intent param '{k}' has a URL (move to #ref): {v}")
            elif re.search(r"[a-fA-F0-9]{12,}", v):
                _emit(res, enforce, f"V1: intent param '{k}' looks like a hash (move to #ref): {v}")
            elif k not in _NUMERIC_INTENT_KEYS and re.search(r"\d", v):
                _emit(res, enforce, f"V1: intent param '{k}' has a number (consider #fact): {k}={v}")


def _v2_evidence(msg: Message, res: Result) -> None:
    for r in msg.records:
        if r.kind != "evid":
            continue
        missing = [k for k in ("claim", "src", "conf") if k not in r.kv]
        if missing:
            res.errors.append(f"V2: #evid missing {', '.join(missing)}: {r.raw}")
            continue
        try:
            c = float(r.kv["conf"])
            if not (0.0 <= c <= 1.0):
                res.errors.append(f"V2: #evid conf must be in [0,1]: {r.raw}")
        except ValueError:
            res.errors.append(f"V2: #evid conf is not a number: {r.raw}")


def _v3_refs(msg: Message, res: Result) -> None:
    for r in msg.records:
        if r.kind == "ref":
            for v in r.kv.values():
                if not (is_valid_ref(v) or is_sloc_ref(v)):
                    res.errors.append(f"V3: invalid ref format: {v}")
    deps = msg.headers.get("deps")
    if deps:
        for d in deps.split(","):
            d = d.strip()
            if not d or not (is_valid_ref(d) or is_sloc_ref(d) or _BARE_DEP.match(d)):
                res.errors.append(f"V3: invalid @deps entry: {d!r}")
    for k in ("p", "root", "parent"):
        v = msg.headers.get(k)
        if v and v.startswith("ref:") and not is_valid_ref(v):
            res.errors.append(f"V3: invalid ref in @{k}: {v}")


def _v6_rid_unique(msg: Message, res: Result) -> None:
    seen: set[str] = set()
    for r in msg.records:
        if r.rid:
            low = r.rid.lower()
            if low in seen:
                res.errors.append(f"V6: duplicate @rid: {r.rid}")
            seen.add(low)


def _v8_budget(msg: Message, res: Result) -> None:
    b = msg.headers.get("budget")
    if not b:
        return
    bm = _BUDGET.match(b)
    if not bm:
        res.errors.append(f"V8: invalid @budget format: {b}")
        return
    limit, cur = float(bm.group(1)), bm.group(2)
    total = 0.0
    for r in msg.records:
        if r.kind == "cost" and r.kv.get("cur") == cur:
            try:
                total += float(r.kv.get("val", "0"))
            except ValueError:
                pass
    if total > limit:
        res.errors.append(f"V8: total cost {total} {cur} exceeds budget {limit} {cur}")


def _v9_tool_chain(msg: Message, res: Result, strict: bool) -> None:
    call_rids = {r.rid.lower() for r in msg.records if r.kind == "call" and r.rid}
    for r in msg.records:
        if r.kind == "call" and "tool" not in r.kv:
            res.errors.append(f"V9: #call missing 'tool': {r.raw}")
        elif r.kind == "ret":
            if "call" not in r.kv:
                res.errors.append(f"V9: #ret missing 'call': {r.raw}")
            elif r.kv["call"].lower() not in call_rids:
                _emit(res, strict, f"V9: #ret references unknown call '{r.kv['call']}': {r.raw}")
            status = r.kv.get("status")
            if status is None:
                res.errors.append(f"V9: #ret missing 'status': {r.raw}")
            elif status not in ("ok", "err"):
                res.errors.append(f"V9: #ret status must be ok|err: {r.raw}")
        elif r.kind == "think" and "summary" not in r.kv:
            res.errors.append(f"V9: #think missing 'summary': {r.raw}")
        elif r.kind == "edit":
            if "file" not in r.kv:
                res.errors.append(f"V9: #edit missing 'file': {r.raw}")
            ch = r.kv.get("changes")
            if ch is None or not ch.isdigit() or int(ch) < 1:
                res.errors.append(f"V9: #edit 'changes' must be a positive integer: {r.raw}")


def _v11_turn_markers(msg: Message, res: Result) -> None:
    ids: set[str] = set()
    refs: list[tuple[str, str]] = []
    for r in msg.records:
        if r.kind == "marker":
            if r.name in ids:
                res.errors.append(f"V11: duplicate turn marker {r.name}")
            ids.add(r.name)
            if r.parent and r.parent != "-":
                refs.append((r.parent, "parent"))
        if r.m:
            refs.append((r.m, "@m"))
    if ids:
        for ref_id, kind in refs:
            if ref_id not in ids:
                res.errors.append(f"V11: {kind}={ref_id} references undeclared turn marker")


def _v12_columnar(msg: Message, res: Result) -> None:
    for blk in msg.blocks:
        label = f"#{blk.rtype}[{','.join(blk.columns)}]"
        if blk.rtype in COLUMNAR_FORBIDDEN:
            res.errors.append(f"V12: columnar form not allowed for #{blk.rtype} (key is data): {label}")
        if not blk.columns or any(not re.fullmatch(r"[a-z][a-z0-9_]*", c) for c in blk.columns):
            res.errors.append(f"V12: malformed column list: {label}")
            continue
        if len(set(blk.columns)) != len(blk.columns):
            res.errors.append(f"V12: duplicate column key in {label}")
        if not blk.rows:
            res.warnings.append(f"V12: columnar block has no rows: {label}")
        for row in blk.rows:
            if len(row) != len(blk.columns):
                vals = " ".join(('"%s"' % v if q else v) for v, q in row)
                res.errors.append(
                    f"V12: row has {len(row)} field(s), expected {len(blk.columns)} for {label}: {vals}"
                )


def _emit(res: Result, as_error: bool, msg: str) -> None:
    (res.errors if as_error else res.warnings).append(msg)

File: impl/typescript/src/validate.ts
/** PAIRL v1.5 validation rules (V1–V12). */

import { COLUMNAR_FORBIDDEN, type Message, type PairlRecord } from "./model.js";

const REF = /^ref:[A-Za-z0-9_-]+:[^\s]+$/;
const SLOC = /^@[A-Za-z0-9]{1,8}(?:#[A-Za-z0-9_-]{1,8})?$/;
const BARE_DEP = /^[A-Za-z0-9]{1,8}(?:#[A-Za-z0-9_-]{1,8})?$/;
const BUDGET = /^([0-9]+(?:\.[0-9]+)?)([A-Za-z]{1,16})$/;
const COL_KEY = /^[a-z][a-z0-9_]*$/;
const NUMERIC_INTENT_KEYS = new Set(["l", "m"]);

export interface Result {
  valid: boolean;
  errors: string[];
  warnings: string[];
}

export function isValidRef(v: string): boolean {
  if (!v.startsWith("ref:") || v.includes(" ")) return false;
  const main = v.split("#")[0];
  return main.split(":").length >= 3 && REF.test(main);
}
export function isSlocRef(v: string): boolean {
  return SLOC.test(v);
}

export function validate(msg: Message, strict = false): Result {
  const errors: string[] = [...msg.errors];
  const warnings: string[] = [];
  const emit = (asErr: boolean, m: string) => (asErr ? errors : warnings).push(m);

  for (const h of ["v", "ts"]) if (!(h in msg.headers)) errors.push(`missing required header: @${h}`);
  if (!("id" in msg.headers) && !("mid" in msg.headers))
    errors.push("missing required header: @id (or legacy @mid)");

  const hasRule = (name: string) =>
    msg.records.some((r) => r.kind === "rule" && r.kv[name] === "true");

  // V1 — no new facts in intents (policy/heuristic)
  const enforceV1 = strict && hasRule("no_new_facts");
  for (const r of msg.records) {
    if (r.kind !== "intent") continue;
    for (const [k, v] of Object.entries(r.kv)) {
      if (/https?:\/\//.test(v)) emit(enforceV1, `V1: intent param '${k}' has a URL (move to #ref): ${v}`);
      else if (/[a-fA-F0-9]{12,}/.test(v)) emit(enforceV1, `V1: intent param '${k}' looks like a hash (move to #ref): ${v}`);
      else if (!NUMERIC_INTENT_KEYS.has(k) && /\d/.test(v)) emit(enforceV1, `V1: intent param '${k}' has a number (consider #fact): ${k}=${v}`);
    }
  }

  // V2 — evidence completeness
  for (const r of msg.records) {
    if (r.kind !== "evid") continue;
    const missing = ["claim", "src", "conf"].filter((k) => !(k in r.kv));
    if (missing.length) { errors.push(`V2: #evid missing ${missing.join(", ")}: ${r.raw}`); continue; }
    const c = Number(r.kv.conf);
    if (Number.isNaN(c)) errors.push(`V2: #evid conf is not a number: ${r.raw}`);
    else if (c < 0 || c > 1) errors.push(`V2: #evid conf must be in [0,1]: ${r.raw}`);
  }

  // V3 — ref format
  for (const r of msg.records) {
    if (r.kind === "ref")
      for (const v of Object.values(r.kv))
        if (!(isValidRef(v) || isSlocRef(v))) errors.push(`V3: invalid ref format: ${v}`);
  }
  if (msg.headers.deps)
    for (const d of msg.headers.deps.split(",").map((x) => x.trim()))
      if (!d || !(isValidRef(d) || isSlocRef(d) || BARE_DEP.test(d)))
        errors.push(`V3: invalid @deps entry: ${JSON.stringify(d)}`);

  // V6 — RID uniqueness
  const seen = new Set<string>();
  for (const r of msg.records) {
    if (!r.rid) continue;
    const low = r.rid.toLowerCase();
    if (seen.has(low)) errors.push(`V6: duplicate @rid: ${r.rid}`);
    seen.add(low);
  }

  // V8 — budget compliance
  if (msg.headers.budget) {
    const bm = BUDGET.exec(msg.headers.budget);
    if (!bm) errors.push(`V8: invalid @budget format: ${msg.headers.budget}`);
    else {
      const limit = Number(bm[1]);
      const cur = bm[2];
      let total = 0;
      for (const r of msg.records)
        if (r.kind === "cost" && r.kv.cur === cur) total += Number(r.kv.val ?? "0") || 0;
      if (total > limit) errors.push(`V8: total cost ${total} ${cur} exceeds budget ${limit} ${cur}`);
    }
  }

  // V9 — tool chain integrity
  const callRids = new Set(msg.records.filter((r) => r.kind === "call" && r.rid).map((r) => r.rid!.toLowerCase()));
  for (const r of msg.records) {
    if (r.kind === "call" && !("tool" in r.kv)) errors.push(`V9: #call missing 'tool': ${r.raw}`);
    else if (r.kind === "ret") {
      if (!("call" in r.kv)) errors.push(`V9: #ret missing 'call': ${r.raw}`);
      else if (!callRids.has(r.kv.call.toLowerCase())) emit(strict, `V9: #ret references unknown call '${r.kv.call}': ${r.raw}`);
      if (!("status" in r.kv)) errors.push(`V9: #ret missing 'status': ${r.raw}`);
      else if (r.kv.status !== "ok" && r.kv.status !== "err") errors.push(`V9: #ret status must be ok|err: ${r.raw}`);
    } else if (r.kind === "think" && !("summary" in r.kv)) errors.push(`V9: #think missing 'summary': ${r.raw}`);
    else if (r.kind === "edit") {
      if (!("file" in r.kv)) errors.push(`V9: #edit missing 'file': ${r.raw}`);
      const ch = r.kv.changes;
      if (!ch || !/^\d+$/.test(ch) || Number(ch) < 1) errors.push(`V9: #edit 'changes' must be a positive integer: ${r.raw}`);
    }
  }

  // V11 — turn marker integrity
  const ids = new Set<string>();
  const refs: Array<[string, string]> = [];
  for (const r of msg.records) {
    if (r.kind === "marker") {
      if (r.name && ids.has(r.name)) errors.push(`V11: duplicate turn marker ${r.name}`);
      if (r.name) ids.add(r.name);
      if (r.parent && r.parent !== "-") refs.push([r.parent, "parent"]);
    }
    if (r.m) refs.push([r.m, "@m"]);
  }
  if (ids.size)
    for (const [id, kind] of refs)
      if (!ids.has(id)) errors.push(`V11: ${kind}=${id} references undeclared turn marker`);

  // V12 — columnar block integrity
  for (const blk of msg.blocks) {
    const label = `#${blk.rtype}[${blk.columns.join(",")}]`;
    if (COLUMNAR_FORBIDDEN.has(blk.rtype)) errors.push(`V12: columnar form not allowed for #${blk.rtype} (key is data): ${label}`);
    if (!blk.columns.length || blk.columns.some((c) => !COL_KEY.test(c))) { errors.push(`V12: malformed column list: ${label}`); continue; }
    if (new Set(blk.columns).size !== blk.columns.length) errors.push(`V12: duplicate column key in ${label}`);
    if (!blk.rows.length) warnings.push(`V12: columnar block has no rows: ${label}`);
    for (const row of blk.rows)
      if (row.length !== blk.columns.length)
        errors.push(`V12: row has ${row.length} field(s), expected ${blk.columns.length} for ${label}`);
  }

  return { valid: errors.length === 0, errors, warnings };
}
